### DISCUSSION TITLE
Introducing Repository Pattern

### DISCUSSION THREAD

**Discussion by bentshrimp** [AgainIoT/Open-Set-Go#34] [Feature Suggestion / Share Ideas]
There a design pattern, called 'repository pattern'. Let's discuss whether we should use repository pattern on this project.

# What is Repository Pattern?
Repository pattern allows us to isolate data layer from rest of the application.
![image](https://github.com/AgainIoT/Open-Set-Go/assets/39232867/ee28d38f-705b-42a9-8434-abc0d4a5418e)
Repository pattern capsulize the data layer just like the picture above. Repository sends queries to data sources or map data newly so that it can be used by other domains.

# Advantage
1. Data logic and business logic can be separated.
2. Domain can request data through constant interface.
3. Data source's data can be capsulized
4. Validation through unit test is possible.
5. Solidity between objects decreases.
6. Fit in any design of application.

# Disadvantage
1. Adds an extra layer of abstraction which increases the project's complexity and maintenance cost.
2. Maintaining ORM mappings by hand requires extra work and code.

Here are blogs with very good explanations.
[https://4z7l.github.io/2020/11/24/repository-pattern.html](https://4z7l.github.io/2020/11/24/repository-pattern.html)
[https://www.cosmicpython.com/book/chapter_02_repository.html](https://www.cosmicpython.com/book/chapter_02_repository.html)

**Comment by ymw0407** [comment]
I think it's good! Even if it takes a long time to apply the pattern, I'll try to utilize it! :smile: Thank you for your suggestionAh~
좋은 것 같습니다! 비록 해당 패턴을 적용시키는 것이 오래걸리지라도 활용해보기 위해 노력해보겠습니다! :smile: 제안해주셔서 감사합니다아~
